internal/kafka/internal/services/quota/ams_quota_service_test.go

```diff
@@ -1000,6 +1001,195 @@ func Test_AMSCheckQuota(t *testing.T) {
+					},
+					ClusterAuthorizationFunc: func(cb *v1.ClusterAuthorizationRequest) (*v1.ClusterAuthorizationResponse, error) {
+						sub := v1.SubscriptionBuilder{}
+						sub.ID(cb.ClusterID() + "subscription")
+						sub.Status("Active")
+						ca, _ := v1.NewClusterAuthorizationResponse().Allowed(true).Subscription(&sub).Build()
+						return ca, nil
+					},
+				},
+			},
+			args: args{
+				kafkaID:                         "cluster-7",
+				kafkaInstanceType:               types.STANDARD,
+				kafkaRequestDesiredBillingModel: "STANDARD",
+
+				wantsReserveCalled: true,
+				wantsProductID:     "RHOSAK",
+			},
+		},
+		{
+			// In this test we try to reserve an EVAL for a cluster that has EVAL assigned
+			name: "Reserve EVAL quota for existing cluster with EVAL subscription",
+			fields: fields{
+				kafkaConfig: amsDefaultKafkaConf,
+				ocmClient: &ocm.ClientMock{
+					FindSubscriptionsFunc: func(query string) ([]*v1.Subscription, error) {
+						return ams.FindSubscriptions(query), nil
+					},
+					GetReservedResourcesBySubscriptionIDFunc: func(subscriptionID string) ([]*v1.ReservedResource, error) {
+						return ams.GetReservedResourcesBySubscriptionID(subscriptionID), nil
+					},
+					GetOrganisationIdFromExternalIdFunc: func(externalId string) (string, error) {
+						return fmt.Sprintf("fake-org-id-%s", externalId), nil
+					},
+					ClusterAuthorizationFunc: func(cb *v1.ClusterAuthorizationRequest) (*v1.ClusterAuthorizationResponse, error) {
+						sub := v1.SubscriptionBuilder{}
+						sub.ID(cb.ClusterID() + "subscription")
+						sub.Status("Active")
+						ca, _ := v1.NewClusterAuthorizationResponse().Allowed(true).Subscription(&sub).Build()
+						return ca, nil
+					},
+				},
+			},
+			args: args{
+				kafkaID:                         "cluster-7",
+				kafkaInstanceType:               types.STANDARD,
+				kafkaRequestDesiredBillingModel: "EVAL",
+
+				wantsReserveCalled: false,
+				wantsProductID:     "RHOSAK",
+			},
+		},
+	}
+
+	for _, testcase := range tests {
+		tt := testcase
+
+		kafka := &dbapi.KafkaRequest{
+			Meta: api.Meta{
+				ID: tt.args.kafkaID,
+			},
+			ClusterID:                tt.args.kafkaID,
+			CloudProvider:            tt.args.kafkaRequestCloudProvider,
+			Owner:                    "owner",
+			SizeId:                   "x1",
+			InstanceType:             string(tt.args.kafkaInstanceType),
+			Marketplace:              tt.args.kafkaRequestMarketplace,
+			BillingCloudAccountId:    tt.args.kafkaRequestBillingCloudAccountID,
+			DesiredKafkaBillingModel: tt.args.kafkaRequestDesiredBillingModel,
+		}
+
+		t.Run(tt.name, func(t *testing.T) {
+			g := gomega.NewWithT(t)
+			factory := NewDefaultQuotaServiceFactory(tt.fields.ocmClient, nil, nil, &tt.fields.kafkaConfig)
+			quotaService, _ := factory.GetQuotaService(api.AMSQuotaType)
+
+			_, err := quotaService.ReserveQuotaIfNotAlreadyReserved(kafka)
+			g.Expect(err).ToNot(gomega.HaveOccurred(), "Expecting error not to happen:", err)
```

Comment:
We could add a test that tests some failure

Reply:
That wouldn't be useful.
The only cases the function can fail are:

1. If there is a configuration error and the instance type of an existing Kafka doesn't exist in the configuration
2. If there is a configuration error and the billing model type of an existing Kafka doesn't exist in the configuration
3. If there are communication errors with AMS
4. If it fails to reserve the quota. This relies on the `ReserveQuota` method, that is already been tested elsewhere

There is no need to test cases 1 to 3.

Reply:
I think a valid test is testing what happens if fails to reserve quota, not because of being a bug in the ReserveQuota happening but because this happening during reconciliation time

Reply:
Yes, but this test is testing the `ReserveQuotaIfNotAlreadyReserved` and in that method, the call to `ReserveQuota` is the last call of the method, so testing it when there is no quota doesn't add value to this test.

Reply:
That is accurate for now, not if the implementation is modified in the future.
I think testing that would be more robust but it is not a big concern now so ok if we don't want to add it now.